### PR TITLE
fix so repo download occurs before install, need to anchor

### PR DESCRIPTION
**Pull request by serenakeating**
see my issue raised for more details

### CODE REVIEW

**Comment by serenakeating** [comment]
@msimonin Have you had a chance to look at this PR, we have issues deploying cassandra without it

**Comment by msimonin** [comment]
Hi @serenakeating,

I'll do it soon.

**Comment by pkallos** [comment]
:+1: i am experiencing this issue when testing with the `tests/Vagrantfile`

**Comment by msimonin** [comment]
Anchoring seems a reasonnable way to do it.  ```contain``` function seems to recent to be used here.
Thanks @serenakeating !

File: manifests/repo.pp [hunk 1]
```diff
@@ -19,6 +19,8 @@
                 key          => $key_id,
                 key_source   => $gpgkey,
                 pin          => $pin,
+                before       => Anchor['cassandra::repo::end'],
```
Comment (msimonin) [inline manifests/repo.pp]: Can't we write :  
```
before => Package['dsc']
```
to force the source list to be updated before the package installation ?

File: manifests/repo.pp [hunk 2]
```diff
@@ -28,10 +30,14 @@
                 gpgkey    => $gpgkey,
                 gpgcheck  => $gpgcheck,
                 enabled   => $enabled,
+                before    => Anchor['cassandra::repo::end'],
```
Comment (msimonin) [inline manifests/repo.pp]: The same here ?

**Comment by yeungda** [comment]
Hi, @msimonin, I've been working with @serenakeating on this change. I hope this comment helps answer your question and allows us to get closer to fixing the ordering issue.

**TLDR;** Both options are fine! I prefer containment.

I'll just try and discuss the two solutions a bit. Hopefully you'll understand the differences and be able to choose.

First, the proposed containment solution.

In `Class['cassandra']`, the code asserts:

    Class['cassandra::repo'] -> Class['cassandra::install']

By containing `Class['cassandra::repo::debian']` within `Class['cassandra::repo']`, we allow the above assertion to work as expected.

This means that in the future, you can add additional assertions about the order of `Class['cassandra::repo']` without worrying about how it works inside. It will be easy to add them to `Class['cassandra']`, so your assertions will be at the top level of your class hierarchy. You can also rename the `cassandra::repo::*` classes, with minimal impact on other classes in your codebase.

Now, for the `before => Package['dsc']` solution:

In `Class['cassandra']`, the code asserts:

    Class['cassandra::repo'] -> Class['cassandra::install']

This assertion isn't actually honoured at all, since all of the resources `Class['cassandra::repo']` includes are uncontained classes.

Without containment, maintainers will probably need to remember not to depend on `Class['cassandra::repo']`.

We then ask `Class['cassandra::repo']` to assert:

    Class['cassandra::repo::debian'] -> Package['dsc']

This introduces a coupling between `Class['cassandra::repo']` and `Class['cassandra:install']`. So it increases the risk that by changing `Class['cassandra:install']` you will break `Class['cassandra:repo']`.
This also means that any future ordering requirements between classes in your codebase will need to be specified in a similar way, increasing coupling further.
It also leaves you with two different ways of asserting that repo should happen before install, one of which that doesn't work, one of which does.

If you go for this, I would suggest that you remove the assertion:

    Class['cassandra::repo'] -> Class['cassandra::install']

Having said all that, which option would you prefer to merge? We'll change the Pull Request to suit.

**Comment by msimonin** [comment]
@yeungda you convinced me :)

**Comment by msimonin** [comment]
Btw I'm interesting in using pythos. Is there a manifest that I can use ?

**Comment by serenakeating** [comment]
As far as I know there is not a public one, we ended up writing our own. Though not sure if we can opensource it as it was for the company I work for. There are some manifests in the excscale github project but they are more for cassandra.

It was very simple though, install the package, run the service and configure the /etc/pithos/pithos.yaml via an pithos.yaml.erb file which I just copied and parameterised.

The only extra thing I had to do was initalise the schemas:
  exec {'pithos initialisation':
      command => "pithos -a install-schema",
      path    => '/bin:/usr/bin',
      require => File['/etc/pithos/pithos.yaml'],
      refreshonly => true,
  }
